### PR TITLE
[Security] Cast user to string before calling loadUserByUsername()

### LINKED ISSUE #36054: PreAuthenticatedAuthenticationProvider calls loadUserByUsername with a user object rather than a string
**Symfony version(s) affected**: 4.4.5 (likely others as well)

**Description**  
I have a custom User provider, based on the one provided in `symfony/ldap`. I encountered an exception due to the fact that the provider's `loadUserByUsername` method was passed a user object rather than a string.

Reading the stack trace, I found this snippet in `PreAuthenticatedAuthenticationProvider`:
```php
# in vendor\symfony\security-core\Authentication\Provider\PreAuthenticatedAuthenticationProvider.php (line 57)

        if (!$user = $token->getUser()) {
            throw new BadCredentialsException('No pre-authenticated principal found in request.');
        }
        $user = $this->userProvider->loadUserByUsername($user);
```
If this behavior is expected then the documentation, the PHPDoc block in `UserProviderInterface` and the stub class generated by the Maker bundle all need to be revised, since they are quite clear that that method is going to receive a string.

**How to reproduce**  
This is a rare case of a problem that is evident by mere static code analysis.

The bug was triggered by an `authorizationChecker->isGranted('IS_AUTHENTICATED_REMEMBERED')` call made by `stof/doctrine-extensions-bundle` on a page with no anonymous access (even if there is a third party bundle involved, I highly doubt that a call to `isGranted` could be done wrong). My custom user provider is close enough to the one in `symfony/ldap` that it should also throw an exception.

**Workaround**
Two workarounds are working for me so far:

- implementing `__toString` in my user class
- checking for the type of the `$username` parameter in my user provider and invoking `getUsername` in there.

**Possible Solution**  
If the behaviour is expected, amend the documentation and PHPDoc blocks accordingly. 

If this is indeed a bug, a naive one-line patch could be to introduce a call to `getUsername` in `PreAuthenticatedAuthenticationProvider` but more likely the right fix will have to do with the logic that class is using. Maybe it should call `refreshUser` instead?

### PR DESCRIPTION
**Pull request by chalasr**
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix #36054 
| License       | MIT
| Doc PR        | -

### CODE REVIEW

**Comment by wouterj** [comment]
Hmm, I'm not convinced this fixes any bug (well it fixes the problem mentioned in the issue, but I highly doubt that this is a problem). Let me explain quickly:

* AuthenticationProviders get pre-authenticated tokens. These are equal to the authenticated tokens, with one major difference: they contain the "credentials" hold to login.
* Loading the user is part of what an AuthenticationProvider does. The token passed to `authenticate()` holds e.g. the input of a `_username` form field, the authentication provider loads a user based on this value using a user provider.

For some reason, the application in #36054 is passing a token with user object to `AuthenticationManager::authenticate()`. It just so happends that this user is using a PreAuthenticatedProvider, but any provider is broken if you pass a token with a user object (instead of a string). See e.g. https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Security/Core/Authentication/Provider/UserAuthenticationProvider.php#L64

So what should be investigated here is why and how there already is a user object before `AuthenticationManager#authenticate()` :)

**Comment by nescim** [comment]
> So what should be investigated here is why and how there already is a user object before AuthenticationManager#authenticate() :)

On that count, I can tell that the method `getUserByUsername` in my user provider was called three times in the same http request. The first two times it was called with a string while the third it was called with a user object.

I know because I put a `dump` call in it to debug the issue and I remembe rit because I was surprised that it was called more than once in a single request.

**Comment by chalasr** [comment]
@nescim Could you please add the logs for that request to your [issue](https://github.com/symfony/symfony/issues/36054)? 
Your security config also? Or even better, create a [reproducer](https://symfony.com/doc/current/contributing/code/reproducer.html#reproducing-complex-bugs)?

@wouterj I agree about fixing the root cause. But still, the current code makes assumptions that go against the contract: 
`PreAuthenticatedToken::getUser()`'s return type is `\Stringable|UserInterface|string`.

Blindly passing its return value to a method that expects a string is calling for bugs.
At this point, I think we would better throw a meaningful logic exception to not let the type error happen.
But yea, we need to confirm that there is no legit use case for getting a non-string user at this stage before going this way.

File: src/Symfony/Component/Security/Core/Authentication/Provider/PreAuthenticatedAuthenticationProvider.php [hunk 1]
```diff
@@ -54,7 +55,7 @@ public function authenticate(TokenInterface $token)
             throw new BadCredentialsException('No pre-authenticated principal found in request.');
         }
 
-        $user = $this->userProvider->loadUserByUsername($user);
+        $user = $this->userProvider->loadUserByUsername($user instanceof UserInterface ? $user->getUsername() : (string) $user);
```
Comment (linaori) [inline src/Symfony/Component/Security/Core/Authentication/Provider/PreAuthenticatedAuthenticationProvider.php]: What about using `$token->getUsername()` instead? The code in the Abstract token already does exactly that check 😄  
```suggestion
        $user = $this->userProvider->loadUserByUsername($token->getUsername());
```

**Comment by nicolas-grekas** [comment]
What's the consensus here?

**Comment by nescim** [comment]
> @nescim Could you please add the logs for that request to your [issue](https://github.com/symfony/symfony/issues/36054)?
> Your security config also? Or even better, create a [reproducer](https://symfony.com/doc/current/contributing/code/reproducer.html#reproducing-complex-bugs)?

Unfortunately dedicating time to this is quite hard at this point and among the more serious effects the CoViD19 pandemic had, my free time is now absolutely zero. I'll do what I can when I can but no promises.

I'm not even sure how useful a reproducer would be here. I'd start from the intended behaviour of the offending class (which is not at all clear to me) and double check what is supposed to do in that execution branch.

**Comment by chalasr** [comment]
Closing then, can be reopened if reproduced.

**Pull request closed without merging**